src/Geta.NotFoundHandler.Optimizely/Core/AutomaticRedirects/MovedContentRegistratorBackgroundService.cs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) Geta Digital. All rights reserved.
+// Licensed under Apache-2.0. See the LICENSE file in the project root for more information
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using EPiServer.Core;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+
+namespace Geta.NotFoundHandler.Optimizely.Core.AutomaticRedirects
+{
+    public class MovedContentRegistratorBackgroundService : BackgroundService
+    {
+        private readonly ChannelMovedContentRegistratorQueue _registratorQueue;
+        private readonly ContentKeyGenerator _contentKeyGenerator;
+        private readonly IAutomaticRedirectsService _automaticRedirectsService;
+        private readonly IContentUrlHistoryLoader _contentUrlHistoryLoader;
+        private readonly ContentUrlIndexer _contentUrlIndexer;
+        private readonly ILogger<MovedContentRegistratorBackgroundService> _logger;
+
+        public MovedContentRegistratorBackgroundService(
+            ChannelMovedContentRegistratorQueue registratorQueue,
+            ContentKeyGenerator contentKeyGenerator,
+            IAutomaticRedirectsService automaticRedirectsService,
+            IContentUrlHistoryLoader contentUrlHistoryLoader,
+            ContentUrlIndexer contentUrlIndexer,
+            ILogger<MovedContentRegistratorBackgroundService> logger)
+        {
+            _registratorQueue = registratorQueue;
+            _contentKeyGenerator = contentKeyGenerator;
+            _automaticRedirectsService = automaticRedirectsService;
+            _contentUrlHistoryLoader = contentUrlHistoryLoader;
+            _contentUrlIndexer = contentUrlIndexer;
+            _logger = logger;
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken cancellationToken)
+        {
+            while (await _registratorQueue.WaitToReadAsync(cancellationToken))
```

Comment:
if you expose `Channel.Reader.ReadAllAsync()` you can actually write then async foreach:

```csharp
await foreach (var data in _channel.Reader.ReadAllAsync(stoppingToken)) { ... }
```

a bit shorter compared to double `while`